### PR TITLE
New polarization correction algorithm for reflectometers

### LINKED ISSUE #20962: Polarization corrections for reflectometry instruments at ILL
This issue tracks the progress of polarization corrections for the ILL reflectometry workflow (#20359).

[`PolarizationCorrection`](http://docs.mantidproject.org/nightly/algorithms/PolarizationCorrection-v1.html) does the right thing, but has a few issues with regards to ILL:
- The efficiencies are given as polynomials. At the moment, the D17 instrument uses piecewise linear functions.
- Incomplete experiments are unsupported. For example, in many cases *R(01) = R(10)* and consequently only one flipping option is measured.

One way to overcome the shortcomings of `PolarizationCorrection` would be:
- Add optional/alternative input workspace properties for the efficiencies so it wouldn't matter if polynomials/interpolation/tabulated values were originally used.
- Add an algorithm to generate the polarization efficiency workspaces from whatever file format ILL ends up using.
- Make `PolarizationCorrection` to work with incomplete datasets.

### PR DESCRIPTION
**Pull request by soininen**
This PR introduces new `PolarizationEfficiencyCor` algorithm for reflectometrer polarization analysis. The polarization corrections are based on [A. R. Wildes (2006), Neutron News, 17:2, 17-25].

**To test:**

Scientific validation will be done by scientists at ILL. For the purpose of this PR, try to break the algorithm. The following script could serve as a basis for testing:
```python
# The unit test data dir has to be in Mantid's data search directories.
ws00 = CreateSampleWorkspace(
    Function='Quasielastic',
    NumBanks=1,
    BankPixelWidth=1,
    XUnit='Wavelength',
    XMin=0.1,
    XMax=10.,
    BinWidth=0.1
)
ws01 = CloneWorkspace(ws00)
ws10 = CloneWorkspace(ws00)
ws11 = CloneWorkspace(ws00)

factors = LoadILLPolarizationFactors(
    Filename='ILL/D17/PolarizationFactors.txt',
    WavelengthReference=ws00
)

PolarizationEfficiencyCor(
    Inputworkspaces='ws00, ws01, ws10, ws11',
    OutputWorkspace='corr',
    Efficiencies=factors,
    Flippers='00, 01, 10, 11'
)
```

Fixes #20962.

**Release Notes** 

A note was added to the reflectometry release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/PolarizationEfficiencyCor.cpp [hunk 1]
```diff
@@ -0,0 +1,928 @@
...
+PolarizationEfficiencyCor::WorkspaceMap
+PolarizationEfficiencyCor::fullCorrections(const WorkspaceMap &inputs,
+                                           const EfficiencyMap &efficiencies) {
+  using namespace boost::math;
+  checkInputExists(inputs.mmWS, Flippers::OnOn);
+  checkInputExists(inputs.mpWS, Flippers::OnOff);
+  checkInputExists(inputs.pmWS, Flippers::OffOn);
+  checkInputExists(inputs.ppWS, Flippers::OffOff);
+  WorkspaceMap outputs;
+  // TODO check if history is retained with create().
```
Comment (gemmaguest) [inline Framework/Algorithms/src/PolarizationEfficiencyCor.cpp]: Is this `TODO` left in intentionally?

File: Framework/Algorithms/src/PolarizationEfficiencyCor.cpp [hunk 2]
```diff
@@ -0,0 +1,928 @@
...
+/**
+ * Correct for non-ideal instrument effects.
+ * Perform full polarization corrections. All flipper configurations
+ * (00, 01, 10 and 11) are needed for this.
+ * @param inputs a set of workspaces to correct
+ * @param efficiencies a set of efficiency factors
+ * @return a set of corrected workspaces
+ */
+PolarizationEfficiencyCor::WorkspaceMap
+PolarizationEfficiencyCor::fullCorrections(const WorkspaceMap &inputs,
```
Comment (gemmaguest) [inline Framework/Algorithms/src/PolarizationEfficiencyCor.cpp]: Could this function be split up to reduce the length/nesting?

File: Framework/Algorithms/src/PolarizationEfficiencyCor.cpp [hunk 3]
```diff
@@ -0,0 +1,928 @@
...
+ * Correct for non-ideal instrument effects.
+ * Only 00 and 11 flipper configurations need to be provided;
+ * the missing 01 and 10 data is solved from the assumption that
+ * in the corrected data, R01 = R10 = 0.
+ * @param inputs a set of workspaces to correct
+ * @param efficiencies a set of efficiency factors
+ * @return a set of corrected workspaces
+ */
+PolarizationEfficiencyCor::WorkspaceMap
+PolarizationEfficiencyCor::twoInputCorrections(
```
Comment (gemmaguest) [inline Framework/Algorithms/src/PolarizationEfficiencyCor.cpp]: Could this function be split up to reduce the length/nesting?

**Review by gemmaguest — CHANGES_REQUESTED** [review]
The changes look good and behave as described. I haven't managed to break the algorithm. Gives sensible errors with:
- missing/invalid workspaces
- mismatched sizes or X values for the input workspaces 
- mismatched X values or missing labels in the factors workspace
- an invalid flippers string
- mismatch between number of flippers and input workspaces
:+1:

Just a few minor in-line queries.

**Comment by soininen** [comment]
@gemmaguest I've refactored the methods as requested.

**Review by gemmaguest — APPROVED** [review]

**Comment by gemmaguest** [comment]
Looks good to me now :+1:

**Pull request merged**